### DISCUSSION TITLE
Discussion about development preview of CodeReady Containers

### DISCUSSION THREAD

**Discussion by praveenkumar** [crc-org/crc#2916] [Show and tell]
This is a development preview of CodeReady Containers.
-----

This last month we have worked hard on new functionality: From a new desktop application to the ability to run a Podman environment. We would like to thank you for all the feedback we got over the years. To celebrate we like to share the download of the developer preview with you and join in the discussions to improve it even more.

The new version can be found at https://github.com/code-ready/crc/releases/tag/dev-1.99.1 This includes the new installers for Windows and macOS. Bundles will be downloaded on demand based on the environment you choose. 

Please read the instructions carefully as some things are to ease with testing and to collect feedback. If you experience any issues or have suggesyions, please consider commenting on this discussion. 

----
As part of this developer preview we provide

  * A new tray/desktop application (macOS and Windows only)
  * Allows for running an OpenShift or Podman bundle
 
### Note
The following is part of the development preview and will not be in the final release.

  * Telemetry is enforced for this release to collect issues
  * Onboarding process and setup wizard is restarted on each start of the desktop application. This is because preset can otherwise only be change from the command line. Newer version allow this to be changed from the configuration window. 
  * MacOS package is not signed so user need to manually add this package as exception from `Security & Privacy` system preferences.

**Comment by gbraad** [comment]
With this we have also published a dev preview of a cockpit module that can help with the administration of a remote server running CRC without the need to log in using SSH.

The cockpit module and install instructions can be found at: https://github.com/code-ready/cockpit-crc/releases/tag/dev-0.9.0

**Comment by e-minguez** [comment]
```
~ 
❯ crc version
CodeReady Containers version: 1.99.0+c9e2be01
OpenShift version: 4.9.10 (not embedded in executable)

~ 
❯ crc setup
INFO Using bundle path /home/edu/.crc/cache/crc_libvirt_4.9.10.crcbundle 
INFO Checking if running as non-root              
INFO Checking if running inside WSL2              
INFO Checking if crc-admin-helper executable is cached 
INFO Checking for obsolete admin-helper executable 
INFO Checking if running on a supported CPU architecture 
INFO Checking minimum RAM requirements            
INFO Checking if crc executable symlink exists    
INFO Checking if Virtualization is enabled        
INFO Checking if KVM is enabled                   
INFO Checking if libvirt is installed             
INFO Checking if user is part of libvirt group    
INFO Adding user to libvirt group                 
INFO Using root access: Adding user to the libvirt group 
INFO Checking if active user/process is currently part of the libvirt group 
INFO Checking if libvirt daemon is running        
INFO Checking if a supported libvirt version is installed 
INFO Checking if crc-driver-libvirt is installed  
INFO Checking crc daemon systemd service          
INFO Checking crc daemon systemd socket units     
INFO Checking if systemd-networkd is running      
INFO Checking if NetworkManager is installed      
INFO Checking if NetworkManager service is running 
INFO Checking if dnsmasq configurations file exist for NetworkManager 
INFO Checking if the systemd-resolved service is running 
INFO Checking if /etc/NetworkManager/dispatcher.d/99-crc.sh exists 
INFO Checking if libvirt 'crc' network is available 
INFO Checking if libvirt 'crc' network is active  
INFO Checking if CRC bundle is extracted in '$HOME/.crc' 
INFO Checking if /home/edu/.crc/cache/crc_libvirt_4.9.10.crcbundle exists 
Your system is correctly setup for using CodeReady Containers, you can now run 'crc start' to start the OpenShift cluster

~ 
❯ crc start
INFO Checking if running as non-root              
INFO Checking if running inside WSL2              
INFO Checking if crc-admin-helper executable is cached 
INFO Checking for obsolete admin-helper executable 
INFO Checking if running on a supported CPU architecture 
INFO Checking minimum RAM requirements            
INFO Checking if crc executable symlink exists    
INFO Checking if Virtualization is enabled        
INFO Checking if KVM is enabled                   
INFO Checking if libvirt is installed             
INFO Checking if user is part of libvirt group    
WARN Preflight checks failed during `crc start`, please try to run `crc setup` first in case you haven't done so yet 
Failed to get the groups user 'edu' belongs to

~ 
❯ grep libvirt /etc/group
libvirt:x:983:edu

~ 
❯ cat /etc/fedora-release 
Fedora release 35 (Thirty Five)

~ 
❯ lsb_release -a
LSB Version:	:core-4.1-amd64:core-4.1-noarch
Distributor ID:	Fedora
Description:	Fedora release 35 (Thirty Five)
Release:	35
Codename:	ThirtyFive
```

**Reply by cfergeau** [reply]
After adding the user to the `libvirt` group, you need to logout/login again (or reboot, ...) so that the change is taken into account. If you run `groups` before that, it won't show your user belonging to the `libvirt` group, this might be the issue you are seeing.

**Reply by e-minguez** [reply]
After rebooting:

```
~ 
❯ groups
edu wheel wireshark libvirt

~ 
❯ crc start
INFO Checking if running as non-root              
INFO Checking if running inside WSL2              
INFO Checking if crc-admin-helper executable is cached 
INFO Checking for obsolete admin-helper executable 
INFO Checking if running on a supported CPU architecture 
INFO Checking minimum RAM requirements            
INFO Checking if crc executable symlink exists    
INFO Checking if Virtualization is enabled        
INFO Checking if KVM is enabled                   
INFO Checking if libvirt is installed             
INFO Checking if user is part of libvirt group    
WARN Preflight checks failed during `crc start`, please try to run `crc setup` first in case you haven't done so yet 
Failed to get the groups user 'edu' belongs to

~ 
❯ crc start --log-level=debug
DEBU CodeReady Containers version: 1.99.0+c9e2be01 
DEBU OpenShift version: 4.9.10 (not embedded in executable) 
DEBU Running 'crc start'                          
DEBU Total memory of system is 33403764736 bytes  
DEBU No new version available. The latest version is 1.37.0 
DEBU Checking file: /home/edu/.crc/machines/crc/.crc-exist 
DEBU Checking if systemd-resolved.service is running 
DEBU Running 'systemctl status systemd-resolved.service' 
DEBU systemd-resolved.service is already running  
INFO Checking if running as non-root              
INFO Checking if running inside WSL2              
INFO Checking if crc-admin-helper executable is cached 
DEBU Running '/home/edu/.crc/bin/crc-admin-helper-linux --version' 
DEBU Found crc-admin-helper-linux version v0.0.8  
DEBU crc-admin-helper executable already cached   
INFO Checking for obsolete admin-helper executable 
DEBU Checking if an older admin-helper executable is installed 
DEBU No older admin-helper executable found       
INFO Checking if running on a supported CPU architecture 
INFO Checking minimum RAM requirements            
DEBU Total memory of system is 33403764736 bytes  
INFO Checking if crc executable symlink exists    
INFO Checking if Virtualization is enabled        
DEBU Checking if the vmx/svm flags are present in /proc/cpuinfo 
DEBU CPU virtualization flags are good            
INFO Checking if KVM is enabled                   
DEBU Checking if /dev/kvm exists                  
DEBU /dev/kvm was found                           
INFO Checking if libvirt is installed             
DEBU Checking if 'virsh' is available             
DEBU 'virsh' was found in /usr/bin/virsh          
DEBU Checking 'virsh capabilities' for libvirtd/qemu availability 
DEBU Running 'virsh --readonly --connect qemu:///system capabilities' 
DEBU Found x86_64 hypervisor with 'hvm' capabilities 
INFO Checking if user is part of libvirt group    
DEBU Checking if current user is part of the libvirt group 
DEBU Failed to get the groups user 'edu' belongs to 
WARN Preflight checks failed during `crc start`, please try to run `crc setup` first in case you haven't done so yet 
Failed to get the groups user 'edu' belongs to
```

**Reply by cfergeau** [reply]
I'm also seeing this with the binary I downloaded from github, I'll investigate in the coming days. It should be possible to workaround this with:
```
crc config set skip-check-user-in-libvirt-group true
```

**Reply by cfergeau** [reply]
I'm thinking the binary on github was built without cgo (cross-built?), which would explain this failure:
https://cs.opensource.google/go/go/+/refs/tags/go1.17.5:src/os/user/lookup_stubs.go
```
func listGroups(*User) ([]string, error) {
	if runtime.GOOS == "android" || runtime.GOOS == "aix" {
		return nil, fmt.Errorf("user: GroupIds not implemented on %s", runtime.GOOS)
	}
	return nil, errors.New("user: GroupIds requires cgo")
}
```

**Reply by gbraad** [reply]
This got recorded and reproduced: https://github.com/code-ready/crc/issues/2922 We have fixed the issue and replaced the release file for Linux.

**Comment by bentito** [comment]
Sorry, but what's the procedure (on MacOS) for switching from say `1.36.0+c0f4e0d3` version to this dev preview version? I did `crc delete`, ran the pkg installer, am running `crc setup` now, but I'm pretty sure it didn't take...

**Reply by bentito** [reply]
FWIW: I was able to transition to the dev preview by clobbering all older CRC stuff, `crc delete`, `rm -rf /Applications/CodeReady*`, `rm -rf ~/.crc`. Then ran the pkg installer again and upon launch of CodeReady app was presented with the installer setup choice of OCP vs. Podman. Not sure if this brute force approach was really needed or not...

I am trying out the Podman option because that is what I am curious about...

**Comment by bentito** [comment]
Oof, I made the mistake of clicking one of those buttons ("Example Deployments" in this case). That changed the installer progress view to some marketing page with no way back to viewing progress. I am blind on the install progress now...

**Reply by gbraad** [reply]
This has been reported in fixed for future releases: https://github.com/code-ready/tray-electron/issues/52

**Comment by jeffmaury** [comment]
We should look into 2 different and parallels preset as I'm looking for a replacement for Docker Desktop
- I cannot use the OpenShift preset for container use as I had to delete this env on a regular basis because it is not stable enough based on the manipulations done on the cluster

**Reply by gbraad** [reply]
The OpenShift bundle is based on RHCOS and also has podman, but currently does not have same stability as it is not the usecase for this platform. This is why we added the podman option. Currently the options aren't possible at the same, and there is plan to change this. However, based on feedback we might reconsider. We will keep an eye on issues and the telemetry stats. Alternatively, consider the podman machine option?

**Comment by jeffmaury** [comment]
Also tried to export/import images from Docker Desktop to podman present without success:
- `podman run -it --rm myimportedimage` gives:
```
Error: no command or entrypoint provided, and no CMD or ENTRYPOINT from image
```
- `podman run -it --rm myimportedimage bash` gives:
```
Error: error preparing container 2de9e909fbd08a1cbe5b8e85a262eefadf26774346ad6c6d024a60399f113805 for attach: executable file `bash` not found in $PATH: No such file or directory: OCI runtime attempted to invoke a command that was not found
```

**Reply by gbraad** [reply]
Can you share more about the image? I do not have this issue.
Perhaps better to file an issue with podman too.

**Reply by jeffmaury** [reply]
That's difficult because it is based on one of my day 2 day images but I will try to reproduce when back from PTO

**Reply by gbraad** [reply]
We will run some tests too. Else find @vrothberg on here or Slack (pinged him internally).

**Reply by jeffmaury** [reply]
Here are the steps to reproduce:
- `docker image save jbosstools/jbosstools-devdoc -o jbosstools-devdoc.tar`
- `podman image import jbosstools-devdoc.tar jbosstools-devdoc`
- `podman run -it --rm jbosstools-devdoc`

Also got the following error:
```
podman run -it --rm jbosstools/jbosstools-devdoc
Error: short-name resolution enforced but cannot prompt without a TTY
```

**Reply by jeffmaury** [reply]
Forget about this issue. load is to be used and not import. The only remaining problem is now regarding the TTY

**Reply by jeffmaury** [reply]
The TTY issue seems to be caused by the terminal software (ConEmu) I'm using. If running from cmd.exe directly, it is working as expected

**Reply by vrothberg** [reply]
Thank you for pulling me in.

> The TTY issue seems to be caused by the terminal software (ConEmu) I'm using. If running from cmd.exe directly, it is working as expected

That is very curious. This happens when stdout or stdin are not a proper TTY; probably an issue in ConEmu.

**Reply by gbraad** [reply]
It is... Windows Terminal doesn't have this problem

**Comment by jeffmaury** [comment]
Just installed latest development 1.99.1. The installer ran fine but as I had a previous version of CRC installed, it stopped the tray but then nothing happened. Tried to start crc but got a missing bundle message. After I reboot the machine, I saw the wizard to setup crc and everything was fine.
